compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -1310,7 +1310,10 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                     },
                 };
                 let mut abi = Abi::Aggregate { sized: true };
-                if tag.value.size(dl) == size {
+
+                // Without latter check aligned enums with custom discriminant values
+                // Would result in ICE see the issue #92464 for more info
+                if tag.value.size(dl) == size || variants.iter().all(|layout| layout.is_empty()) {
```

Comment:
what is `size` and `tag.value.size(dl)` in the ICEing example vs without `align(8)`? Can you check why we end up with different values here? I think we should address the issue somewhere earlier

Reply:
The only difference between those two is one with `align(8)` size is 8, latter one's size is 1.

It's been a while since I worked on this but, IIRC the issue was in this enum there are only custom discriminant values, weirdly enough `variants` in these code doesn't have a layout (maybe because it doesn't have sized field ?). So `if` check doesn't trigger because `tag.value.size(dl)` is 1 and `size` is 8. Coincidently one without `align` has `size` 1 and its `tag.value.size(dl)` is 1 it passes the check and runs without any problem. 

I tried looking elsewhere to find why `align` doesn't change the size in the first place but I couldn't, it's way above my expertise, and this was already a hard issue for my second contribution 😺 
```rust
#[repr(align(8))]
enum Aligned {
    Zero = 0,
    One = 1,
}
```

Reply:
Thanks! that analysis makes sense to me.

I can explain the size change: We require all types to be at least as large as their alignment so that when you allocate an array, all elements are right after each other without space between the elements (as that would complicate computing the address of an array element).

From this information the change lgtm, but please squash all commits and rebase over the master branch

Reply:
I think I managed to squash and rebase (not very familiar with git), let me know if I didn't.

Reply:
Doesn't look like it 😅 

I think this should work:

```sh
git fetch origin
git rebase -i origin/master
# the above opens a file, replace all `pick` with `fixup` except for the first commit
git push --force-with-lease
```

Reply:
```
git rebase -i origin/master
# this command didn't show anything except `noop` instead I used git rebase -i HEAD~6
```

Reply:
Oh ^^ I forgot to mention to replace `origin` with the name you gave the rustc repository ^^

But everything looks good now!